### DISCUSSION TITLE
SSR / SSG with mobx-state-tree causes "cannot update component X while another component is being updated" errors.

### DISCUSSION THREAD

**Discussion by alaminut** [vercel/next.js#22111] [Help] [answered]
**What version of Next.js are you using?**

10.0.6

**What version of Node.js are you using?**

14.15.4

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS, Windows

**How are you deploying your application?**

next dev

**Describe the Bug**

I have followed mobx-state-tree example in this link https://github.com/vercel/next.js/tree/canary/examples/with-mobx-state-tree-typescript to integrate mst into my application.

When I run the app, the initial redirect from "index" page to another route triggers "cannot update component X while updating component Y" error. If I directly navigate to the "[lang]/index" route (SSG), then it shows "text mismatch" between server and client.

I'm trying to implement a multi lingual website with nextjs. So my app flow is like the following:

* pages/index is just a redirect page based on user's browser language.
* pages/[lang]/index.js is the actual localized index page. This is a SSG page which demonstrates hydrating store from server fails.
* components/layout.js is responsible to handle route changes and display a global "loading" indicator. This component is an observer as it fires the mobx actions.

**Expected Behavior**

I expected values that come from server side store to be consistent with the client side store. I'm not sure how hydration does not work even though I followed the example.

I also expect not to have "can not update component X while another component is being rendered" errors.

**To Reproduce**

Please follow this minimal reproduction repository: https://github.com/alaminut/next-mobx-bug-reproduction

**Comment by nagaozen** [comment, marked as answer]
This bug also exists in all mobx examples:
* with-mobx
* with-mobx-react-lite
* with-mobx-state-tree
* with-recoil
and so on...

It happens because `store` needs to be initialized, but neither going with `useStore` (`useMemo` backed) nor StoreProvider (with `createContext`) seems to fix this issue.

**Reply by nagaozen** [reply]
By changing `_app.js` to:

```js
import App from 'next/app'
import { Provider } from 'mobx-react'
import { initializeStore } from '../store'

export default class MyApp extends App {
  constructor (props) {
    super(props)
    this.state = {
      store: initializeStore(props.initialState)
    }
  }

  render () {
    const { Component, pageProps } = this.props
    const store = this.state.store
    return (
      <Provider store={store}>
        <Component {...pageProps} />
      </Provider>
    )
  }
}
```

Seems to fix the issue, at least on `with-mobx` example...

**Reply by nagaozen** [reply]
Also, in `with-mobx-react-lite`, it's possible to fix by using this `StoreProvider.js`:

```js
import { createContext, useContext, useState, useEffect } from 'react'
import { Store } from '../store'

export const StoreContext = createContext()

export function StoreProvider ({ children, initialState: initialData }) {
  const [store, setStore] = useState(new Store())
  useEffect(() => {
    if (initialData) {
      store.hydrate(initialData)
    }
  }, [])
  return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
}

export function useStore () {
  const context = useContext(StoreContext)
  if (context === undefined) {
    throw new Error('useStore must be used within StoreProvider')
  }
  return context
}
```

Note that `pages/ssg.js` and `pages/ssr.js` are being initialized without using `initializeStore`...

**Reply by nagaozen** [reply]
Finally, on `with-mobx-state-tree`, it's fixable with the following `pages/_app.js`

```js
import { useState, useEffect } from 'react'
import { Provider } from 'mobx-react'
import { applySnapshot } from 'mobx-state-tree'
import { initializeStore } from '../store'

export default function App ({ Component, pageProps }) {
  const [store, setStore] = useState(initializeStore())
  useEffect(() => {
    if (pageProps.initialState) {
      applySnapshot(store, pageProps.initialState)
    }
  })
  return (
    <Provider store={store}>
      <Component {...pageProps} />
    </Provider>
  )
}
```

**Reply by balfouri1210** [reply]
👍

**Comment by mawaisismail** [comment]
is anyone integrate next. js 14 with mobx ? if yes please let me know
